README.txt
= ruby_picasa

* http://github.com/pangloss/ruby_picasa

== DESCRIPTION:

Provides a super easy to use object layer for authenticating and accessing
Picasa through their API.

== FEATURES:

* Simplifies the process of obtaining both a temporary and a permanent AuthSub
  token.
* Very easy to use API.
* Allows access to both public and private User, Album and Photo data.
* Uses Objectify::Xml to define the XML object-relational layer with a very
  easy to understand DSL. See www.github.com/pangloss/objectify_xml

== PROBLEMS:

* None known.

== SYNOPSIS:

  # 1. Authorize application for access (in a rails controller)
  #
  redirect_to RubyPicasa.authorization_url(auth_result_url)

  # 2. Extract the Picasa token from the request Picasa sends back to your app
  #    and create a permanent AuthSub token. Returns an initialized Picasa
  #    session.  (Called from the Rails action for auth_result_url above)
  picasa = RubyPicasa.authorize_request(self.request)

  # 3. Access the data you are interested in
  @album = picasa.user.albums.first
  @photo = @album.photos.first

  # 4. Display your photos
  image_tag @photo.url
  image_tag @photo.url('160c') # Picasa thumbnail names are predefined

== REQUIREMENTS:

* objectify_xml

== INSTALL:

* gem install ruby-picasa
* gem install pangloss-ruby-picasa --source http://gems.github.com

== LICENSE:

(The MIT License)

Copyright (c) 2009 Darrick Wiebe

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.2.0 / 2009-02-25

* First public release.

=== 0.1.0 / 2009-02-21

* First working version.

=== 0.0.0 / 2009-02-21

* Needed to access Picasa.


spec/spec_helper.rb
require 'rubygems'
require File.expand_path(File.join(File.dirname(__FILE__), '../lib/ruby_picasa'))
require 'spec'
require 'mocha'

def open_file(name)
  open(File.join(File.dirname(__FILE__), File.join('sample', name)))
end

Spec::Runner.configure do |config|
  config.mock_with :mocha
end


spec/ruby_picasa_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')

class Picasa
  class << self
    public :parse_url
  end
  public :auth_header, :with_cache, :class_from_xml, :xml_data
end

describe 'Picasa class methods' do
  it 'should generate an authorization_url' do
    return_url = 'http://example.com/example?example=ex'
    url = Picasa.authorization_url(return_url)
    url.should include(CGI.escape(return_url))
    url.should match(/session=1/)
  end

  describe 'token_from_request' do
    it 'should pluck the token from the request' do
      request = mock('request', :params => { 'token' => 'abc' })
      Picasa.token_from_request(request).should == 'abc'
    end
    it 'should raise if no token is present' do
      request = mock('request', :params => { })
      lambda do
        Picasa.token_from_request(request)
      end.should raise_error(RubyPicasa::PicasaTokenError)
    end
  end

  it 'should authorize a request' do
    Picasa.expects(:token_from_request).with(:request).returns('abc')
    picasa = mock('picasa')
    Picasa.expects(:new).with('abc').returns(picasa)
    picasa.expects(:authorize_token!).with()
    Picasa.authorize_request(:request).should == picasa
  end

  it 'should recognize absolute urls' do
    Picasa.is_url?('http://something.com').should be_true
    Picasa.is_url?('https://something.com').should be_true
    Picasa.is_url?('12323412341').should_not be_true
  end

  describe 'path' do
    it 'should use parse_url and add options' do
      Picasa.expects(:parse_url).with({}).returns(['url', {'a' => 'b'}])
      Picasa.path({}).should ==
        "url?a=b"
    end
    it 'should build the url from user_id and album_id and add options' do
      hash = { :user_id => '123', :album_id => '321' }
      Picasa.expects(:parse_url).with(hash).returns([nil, {}])
      Picasa.path(hash).should ==
        "/data/feed/api/user/123/albumid/321?kind=photo"
    end
    it 'should build the url from special user_id all' do
      hash = { :user_id => 'all' }
      Picasa.expects(:parse_url).with(hash).returns([nil, {}])
      Picasa.path(hash).should ==
        "/data/feed/api/all"
    end
    [ :max_results, :start_index, :tag, :q, :kind,
      :access, :bbox, :l].each do |arg|
      it "should add #{ arg } to options" do
        Picasa.path(:url => 'url', arg => '!value').should ==
          "url?#{ arg.to_s.dasherize }=%21value"
      end
    end
    [ :imgmax, :thumbsize ].each do |arg|
      it "should raise PicasaError with invalid #{ arg } option" do
        lambda do
          Picasa.path(:url => 'url', arg => 'invalid')
        end.should raise_error(RubyPicasa::PicasaError)
      end
    end
    [ :imgmax, :thumbsize ].each do |arg|
      it "should add #{ arg } to options" do
        Picasa.path(:url => 'url', arg => '72').should ==
          "url?#{ arg.to_s.dasherize }=72"
      end
    end
    it 'should ignore unknown options' do
      Picasa.path(:url => 'place', :eggs => 'over_easy').should == 'place'
    end
  end

  describe 'parse_url' do
    it 'should prefer url' do
      hash = { :url => 'url', :user_id => 'user_id', :album_id => 'album_id' }
      Picasa.parse_url(hash).should == ['url', {}]
    end
    it 'should next prefer user_id' do
      Picasa.stubs(:is_url?).returns true
      hash = { :user_id => 'user_id', :album_id => 'album_id' }
      Picasa.parse_url(hash).should == ['user_id', {}]
    end
    it 'should use album_id' do
      Picasa.stubs(:is_url?).returns true
      hash = { :album_id => 'album_id' }
      Picasa.parse_url(hash).should == ['album_id', {}]
    end
    it 'should split up the params' do
      hash = { :url => 'url?specs=fun%21' }
      Picasa.parse_url(hash).should == ['url', { 'specs' => 'fun!' }]
    end
    it 'should not use non-url user_id or album_id' do
      hash = { :user_id => 'user_id', :album_id => 'album_id' }
      Picasa.parse_url(hash).should == [nil, {}]
    end
    it 'should handle with no relevant options' do
      hash = { :saoetu => 'aeu' }
      Picasa.parse_url(hash).should == [nil, {}]
    end
  end
end

describe Picasa do
  def body(text)
    #open_file('user_feed.atom').read
    @response.stubs(:body).returns(text)
  end

  before do
    @response = mock('response')
    @response.stubs(:code).returns '200'
    @http = mock('http')
    @http.stubs(:get).returns @response
    Net::HTTP.stubs(:new).returns(@http)
    @p = Picasa.new 'token'
  end

  it 'should initialize' do
    @p.token.should == 'token'
  end

  describe 'authorize_token!' do
    before do
      @p.expects(:auth_header).returns('Authorization' => 'etc')
      @http.expects(:use_ssl=).with true
      @http.expects(:get).with('/accounts/accounts/AuthSubSessionToken', 
        'Authorization' => 'etc').returns(@response)
    end

    it 'should set the new token' do
      body 'Token=hello'
      @p.authorize_token!
      @p.token.should == 'hello'
    end

    it 'should raise if the token is not found' do
      body 'nothing to see here'
      lambda do
        @p.authorize_token!
      end.should raise_error(RubyPicasa::PicasaTokenError)
      @p.token.should == 'token'
    end
  end

  it 'should get the user' do
    @p.expects(:get).with(:user_id => nil)
    @p.user
  end

  it 'should get an album' do
    @p.expects(:get).with(:album_id => 'album')
    @p.album('album')
  end
  
  it 'should get a url' do
    @p.expects(:get).with(:url => 'the url')
    @p.get_url('the url')
  end

  describe 'search' do
    it 'should prefer given options' do
      @p.expects(:get).with(:q => 'q', :max_results => 20, :user_id => 'me', :kind => 'comments')
      @p.search('q', :max_results => 20, :user_id => 'me', :kind => 'comments', :q => 'wrong')
    end
    it 'should have good defaults' do
      @p.expects(:get).with(:q => 'q', :max_results => 10, :user_id => 'all', :kind => 'photo')
      @p.search('q')
    end
  end

  it 'should get recent photos' do
    @p.expects(:get).with(:recent_photos => true, :max_results => 10)
    @p.recent_photos :max_results => 10
  end

  describe 'album_by_title' do
    before do
      @a1 = mock('a1')
      @a2 = mock('a2')
      @a1.stubs(:title).returns('a1')
      @a2.stubs(:title).returns('a2')
      albums = [ @a1, @a2 ]
      user = mock('user', :albums => albums)
      @p.expects(:user).returns(user)
    end

    it 'should match the title string' do
      @a2.expects(:load).with({}).returns :result
      @p.album_by_title('a2').should == :result
    end

    it 'should match a regex' do
      @a1.expects(:load).with({}).returns :result
      @p.album_by_title(/a\d/).should == :result
    end

    it 'should return nil' do
      @p.album_by_title('zzz').should be_nil
    end
  end

  describe 'xml' do
    it 'should return the body with a 200 status' do
      body 'xml goes here'
      @p.xml.should == 'xml goes here'
    end
    it 'should return nil with a non-200 status' do
      body 'xml goes here'
      @response.expects(:code).returns '404'
      @p.xml.should be_nil
    end
  end

  describe 'get' do
    it 'should call class_from_xml if with_cache yields' do
      @p.expects(:with_cache).with({}).yields(:xml).returns(:result)
      @p.expects(:class_from_xml).with(:xml)
      @p.send(:get).should == :result
    end

    it 'should do nothing if with_cache does not yield' do
      @p.expects(:with_cache).with({}) # doesn't yield
      @p.expects(:class_from_xml).never
      @p.send(:get).should be_nil
    end
  end

  describe 'auth_header' do
    it 'should build an AuthSub header' do
      @p.auth_header.should == { "Authorization" => %{AuthSub token="token"} }
    end

    it 'should do nothing' do
      p = Picasa.new nil
      p.auth_header.should == { }
    end
  end

  describe 'with_cache' do
    it 'yields fresh xml' do
      body 'fresh xml'
      yielded = false
      @p.with_cache(:url => 'place') do |xml|
        yielded = true
        xml.should == 'fresh xml'
      end
      yielded.should be_true
    end

    it 'yields cached xml' do
      @p.instance_variable_get('@request_cache')['place'] = 'some xml'
      yielded = false
      @p.with_cache(:url => 'place') do |xml|
        yielded = true
        xml.should == 'some xml'
      end
      yielded.should be_true
    end
  end

  describe 'xml_data' do
    it 'should extract categories from the xml' do
      xml, feed_schema, entry_schema = @p.xml_data(open_file('album.atom'))
      xml.should be_an_instance_of(Nokogiri::XML::Element)
      feed_schema.should == 'http://schemas.google.com/photos/2007#album'
      entry_schema.should == 'http://schemas.google.com/photos/2007#photo'
    end

    it 'should handle nil' do
      xml, feed_schema, entry_schema = @p.xml_data(nil)
      xml.should be_nil
    end

    it 'should handle bad xml' do
      xml, feed_schema, entry_schema = @p.xml_data('<entry>something went wrong')
      xml.should_not be_nil
      feed_schema.should be_nil
      entry_schema.should be_nil
    end
  end

  describe 'class_from_xml' do
    before do
      @user = 'http://schemas.google.com/photos/2007#user'
      @album = 'http://schemas.google.com/photos/2007#album'
      @photo = 'http://schemas.google.com/photos/2007#photo'
    end

    describe 'valid feed category types' do
      def to_create(klass, feed, entry)
        @object = mock('object', :session= => nil)
        @p.expects(:xml_data).with(:xml).returns([:xml, feed, entry])
        klass.expects(:new).with(:xml, @p).returns(@object)
        @p.class_from_xml(:xml)
      end
      it('user album')  { to_create RubyPicasa::User, @user, @album }
      it('user photo')  { to_create RubyPicasa::RecentPhotos, @user, @photo }
      it('album nil')   { to_create RubyPicasa::Album, @album, nil }
      it('album photo') { to_create RubyPicasa::Album, @album, @photo }
      it('photo nil')   { to_create RubyPicasa::Photo, @photo, nil }
      it('photo photo') { to_create RubyPicasa::Search, @photo, @photo }
    end

    it 'should raise an error for invalid feed category types' do
      @p.expects(:xml_data).with(:xml).returns([:xml, @album, @user])
      lambda do
        @p.class_from_xml(:xml)
      end.should raise_error(RubyPicasa::PicasaError)
    end
  end
end


spec/ruby_picasa/types_spec.rb
require File.join(File.dirname(__FILE__), '../spec_helper')

include RubyPicasa

describe 'a RubyPicasa document', :shared => true do
  it 'should have a feed_id' do
    @object.feed_id.should_not be_nil
  end

  it 'should have an author' do
    unless @no_author
      @object.author.should_not be_nil
      @object.author.name.should == 'Liz'
      @object.author.uri.should == 'http://picasaweb.google.com/liz'
    end
  end

  it 'should get links by name' do
    @object.link('abc').should be_nil
    @object.link('self').href.should_not be_nil
  end

  it 'should do nothing for previous and next' do
    @object.previous.should be_nil if @object.link('previous').nil?
    @object.next.should be_nil if @object.link('next').nil?
  end

  it 'should get the feed' do
    @object.session.expects(:get_url).with(@object.feed_id.
                                            gsub(/entry/, 'feed').
                                            gsub(/default/, 'liz'), {})
    @object.feed
  end

  it 'should have links' do
    @object.links.should_not be_empty
    @object.links.each do |l|
      l.should be_an_instance_of(Objectify::Atom::Link)
    end
  end

  describe 'session' do
    it 'should return @session' do
      @object.session = :sess
      @object.session.should == :sess
    end

    it 'should get the parent session' do
      @object.session = nil
      @parent.expects(:session).returns(:parent_sess)
      @object.session.should == :parent_sess
    end

    it 'should be nil if no parent' do
      @object.session = nil
      @object.expects(:parent).returns nil
      @object.session.should be_nil
    end
  end
end


describe User do
  it_should_behave_like 'a RubyPicasa document'

  before :all do
    @xml = open_file('user.atom').read
  end

  before do
    @parent = mock('parent')
    @object = @user = User.new(@xml, @parent)
    @user.session = mock('session')
  end

  it 'should have albums' do
    @user.albums.length.should == 1
    @user.albums.first.should be_an_instance_of(Album)
  end
end

describe RecentPhotos do
  it_should_behave_like 'a RubyPicasa document'

  before :all do
    @xml = open_file('recent.atom').read
  end

  before do
    @parent = mock('parent')
    @object = @album = RecentPhotos.new(@xml, @parent)
    @album.session = mock('session')
  end

  it 'should have 1 photo' do
    @album.photos.length.should == 1
    @album.photos.first.should be_an_instance_of(Photo)
  end

  it 'should request next' do
    @album.session.expects(:get_url).with('http://picasaweb.google.com/data/feed/api/user/liz?start-index=2&max-results=1&kind=photo').returns(:result)
    @album.next.should == :result
  end

  it 'should not request previous on first page' do
    @album.session.expects(:get_url).never
    @album.previous.should be_nil
  end
end

describe Album do
  it_should_behave_like 'a RubyPicasa document'

  before :all do
    @xml = open_file('album.atom').read
  end

  before do
    @parent = mock('parent')
    @object = @album = Album.new(@xml, @parent)
    @album.session = mock('session')
  end

  it 'should have a numeric id' do
    @object.id.should_not be_nil
    @object.id.to_s.should match(/\A\d+\Z/)
  end

  it 'should have 1 entry' do
    @album.entries.length.should == 1
  end

  it 'should get links by name' do
    @album.link('abc').should be_nil
    @album.link('alternate').href.should == 'http://picasaweb.google.com/liz/Lolcats'
  end

  describe 'photos' do
    it 'should use entries if available' do
      @album.expects(:session).never
      @album.photos.should == @album.entries
    end

    it 'should request photos if needed' do
      @album.entries = []
      new_album = mock('album', :entries => [:photo])
      @album.session.expects(:get_url).with(@album.link(/feed/).href, {}).returns(new_album)
      @album.photos.should == [:photo]
    end
  end

  it 'should be public' do
    @album.public?.should be_true
  end

  it 'should not be private' do
    @album.private?.should be_false
  end

  describe 'first Photo' do
    before do
      @photo = @album.entries.first
      @photo.should be_an_instance_of(Photo)
    end

    it 'should have a parent' do
      @photo.parent.should == @album
    end

    it 'should not have an author' do
      @photo.author.should be_nil
    end

    it 'should have a content' do
      @photo.content.should be_an_instance_of(PhotoUrl)
    end

    it 'should have 3 thumbnails' do
      @photo.thumbnails.length.should == 3
      @photo.thumbnails.each do |t|
        t.should be_an_instance_of(ThumbnailUrl)
      end
    end

    it 'should have a numeric id' do
      @object.id.should_not be_nil
      @object.id.to_s.should match(/\A\d+\Z/)
    end

    it 'should have a default url' do
      @photo.url.should == 'http://lh5.ggpht.com/liz/SKXR5BoXabI/AAAAAAAAAzs/tJQefyM4mFw/invisible_bike.jpg'
    end

    it 'should have thumbnail urls' do
      @photo.url('72').should == 'http://lh5.ggpht.com/liz/SKXR5BoXabI/AAAAAAAAAzs/tJQefyM4mFw/s72/invisible_bike.jpg'
    end

    it 'should have thumbnail info' do
      @photo.thumbnail('72').width.should == 72
    end
  end
end

describe Search do
  it_should_behave_like 'a RubyPicasa document'

  before :all do
    @xml = open_file('search.atom').read
  end

  before do
    @no_author = true
    @parent = mock('parent')
    @object = @search = Search.new(@xml, @parent)
    @search.session = mock('session')
  end

  it 'should have 1 entry' do
    @search.entries.length.should == 1
    @search.entries.first.should be_an_instance_of(Photo)
  end

  it 'should request next' do
    @search.session.expects(:get_url).with('http://picasaweb.google.com/data/feed/api/all?q=puppy&start-index=3&max-results=1').returns(:result)
    @search.next.should == :result
  end

  it 'should request previous' do
    @search.session.expects(:get_url).with('http://picasaweb.google.com/data/feed/api/all?q=puppy&start-index=1&max-results=1').returns(:result)
    @search.previous.should == :result
  end
end



lib/ruby_picasa.rb
require 'objectify_xml'
require 'objectify_xml/atom'
require 'cgi'
require 'net/http'
require 'net/https'
require File.join(File.dirname(__FILE__), 'ruby_picasa/types')

module RubyPicasa
  VERSION = '0.2.1'

  class PicasaError < StandardError
  end

  class PicasaTokenError < PicasaError
  end
end

# == Authorization
#
# RubyPicasa makes authorizing a Rails app easy. It is a two step process: 
#
# First redirect the user to the authorization url, if the user authorizes your
# application, Picasa will redirect the user back to the url you specify (in
# this case authorize_picasa_url). 
#
# Next, pass the Rails request object to the authorize_token method which will
# make the api call to upgrade the token and if successful return an initialized
# Picasa session object ready to use. The token object can be retrieved from the
# token attribute.
#
#   class PicasaController < ApplicationController
#     def request_authorization
#       redirect_to Picasa.authorization_url(authorize_picasa_url)
#     end
#
#     def authorize
#       if Picasa.token_in_request?(request)
#         begin
#           picasa = Picasa.authorize_request(request)
#           current_user.picasa_token = picasa.token
#           current_user.save
#           flash[:notice] = 'Picasa authorization complete'
#           redirect_to picasa_path
#         rescue PicasaTokenError => e
#           # 
#           @error = e.message
#           render
#         end
#       end
#     end
#   end
#
class Picasa
  class << self
    # The user must be redirected to this address to authorize the application
    # to access their Picasa account. The token_from_request and
    # authorize_request methods can be used to handle the resulting redirect
    # from Picasa.
    def authorization_url(return_to_url, request_session = true, secure = false, authsub_url = nil)
      session = request_session ? '1' : '0'
      secure = secure ? '1' : '0'
      return_to_url = CGI.escape(return_to_url)
      url = authsub_url || 'http://www.google.com/accounts/AuthSubRequest'
      "#{ url }?scope=http%3A%2F%2F#{ host }%2Fdata%2F&session=#{ session }&secure=#{ secure }&next=#{ return_to_url }"
    end

    # Takes a Rails request object and extracts the token from it. This would
    # happen in the action that is pointed to by the return_to_url argument
    # when the authorization_url is created.
    def token_from_request(request)
      if token = request.params['token']
        return token
      else
        raise RubyPicasa::PicasaTokenError, 'No Picasa authorization token was found.'
      end
    end

    def token_in_request?(request)
      request.params['token']
    end

    # Takes a Rails request object as in token_from_request, then makes the
    # token authorization request to produce the permanent token. This will
    # only work if request_session was true when you created the
    # authorization_url.
    def authorize_request(request)
      p = Picasa.new(token_from_request(request))
      p.authorize_token!
      p
    end

    # The url to make requests to without the protocol or path.
    def host
      @host ||= 'picasaweb.google.com'
    end

    # In the unlikely event that you need to access this api on a different url,
    # you can set it here. It defaults to picasaweb.google.com
    def host=(h)
      @host = h
    end

    # A simple test used to determine if a given resource id is it's full
    # identifier url. This is not intended to be a general purpose method as the
    # test is just a check for the http/https protocol prefix.
    def is_url?(path)
      path.to_s =~ %r{\Ahttps?://}
    end

    # For more on possible options and their meanings, see: 
    # http://code.google.com/apis/picasaweb/reference.html
    #
    # The following values are valid for the thumbsize and imgmax query
    # parameters and are embeddable on a webpage. These images are available as
    # both cropped(c) and uncropped(u) sizes by appending c or u to the size.
    # As an example, to retrieve a 72 pixel image that is cropped, you would
    # specify 72c, while to retrieve the uncropped image, you would specify 72u
    # for the thumbsize or imgmax query parameter values.
    #
    # 32, 48, 64, 72, 144, 160
    #
    # The following values are valid for the thumbsize and imgmax query
    # parameters and are embeddable on a webpage. These images are available as
    # only uncropped(u) sizes by appending u to the size or just passing the
    # size value without appending anything. 
    #
    # 200, 288, 320, 400, 512, 576, 640, 720, 800
    #
    # The following values are valid for the thumbsize and imgmax query
    # parameters and are not embeddable on a webpage. These image sizes are
    # only available in uncropped format and are accessed using only the size
    # (no u is appended to the size).
    #
    # 912, 1024, 1152, 1280, 1440, 1600
    # 
    def path(args = {})
      path, options = parse_url(args)
      if path.nil?
        path = ["/data/feed/api"]
        if args[:user_id] == 'all'
          path += ["all"]
        else
          path += ["user", CGI.escape(args[:user_id] || 'default')]
        end
        path += ['albumid', CGI.escape(args[:album_id])] if args[:album_id]
        path = path.join('/')
      end
      options['kind'] = 'photo' if args[:recent_photos] or args[:album_id]
      if args[:thumbsize] and not args[:thumbsize].split(/,/).all? { |s| RubyPicasa::Photo::VALID.include?(s) }
        raise RubyPicasa::PicasaError, 'Invalid thumbsize.'
      end
      if args[:imgmax] and not RubyPicasa::Photo::VALID.include?(args[:imgmax])
        raise RubyPicasa::PicasaError, 'Invalid imgmax.'
      end
      [:max_results, :start_index, :tag, :q, :kind,
       :access, :thumbsize, :imgmax, :bbox, :l].each do |arg|
        options[arg.to_s.dasherize] = args[arg] if args[arg]
      end
      if options.empty?
        path
      else
        [path, options.map { |k, v| [k.to_s, CGI.escape(v.to_s)].join('=') }.join('&')].join('?')
      end
    end

    private

    # Extract the path and a hash of key/value pairs from a given url with
    # optional query string.
    def parse_url(args)
      url = args[:url]
      url ||= args[:user_id] if is_url?(args[:user_id]) 
      url ||= args[:album_id] if is_url?(args[:album_id])
      if url
        uri = URI.parse(url)
        path = uri.path
        options = {}
        if uri.query
          uri.query.split('&').each do |query|
            k, v = query.split('=')
            options[k] = CGI.unescape(v)
          end
        end
        [path, options]
      else
        [nil, {}]
      end
    end
  end

  # The AuthSub token currently in use.
  attr_reader :token

  def initialize(token)
    @token = token
    @request_cache = {}
  end

  # Attempt to upgrade the current AuthSub token to a permanent one. This only
  # works if the Picasa session is initialized with a single use token.
  def authorize_token!
    http = Net::HTTP.new("www.google.com", 443)
    http.use_ssl = true
    response = http.get('/accounts/accounts/AuthSubSessionToken', auth_header)
    token = response.body.scan(/Token=(.*)/).flatten.compact.first
    if token
      @token = token
    else
      raise RubyPicasa::PicasaTokenError, 'The request to upgrade to a session token failed.'
    end
    @token
  end

  # Retrieve a RubyPicasa::User record including all user albums.
  def user(user_id_or_url = nil, options = {})
    options = make_options(:user_id, user_id_or_url, options)
    get(options)
  end

  # Retrieve a RubyPicasa::Album record. If you pass an id or a feed url it will
  # include all photos. If you pass an entry url, it will not include photos.
  def album(album_id_or_url, options = {})
    options = make_options(:album_id, album_id_or_url, options)
    get(options)
  end

  # This request does not require authentication. Returns a RubyPicasa::Search
  # object containing the first 10 matches. You can call #next and #previous to
  # navigate the paginated results on the Search object.
  def search(q, options = {})
    h = {}
    h[:max_results] = 10
    h[:user_id] = 'all'
    h[:kind] = 'photo'
    # merge options over h, but merge q over options
    get(h.merge(options).merge(:q => q))
  end

  # Retrieve a RubyPicasa object determined by the type of xml results returned
  # by Picasa. Any supported type of RubyPicasa resource can be requested with
  # this method.
  def get_url(url, options = {})
    options = make_options(:url, url, options)
    get(options)
  end

  # Retrieve a RubyPicasa::RecentPhotos object, essentially a User object which
  # contains photos instead of albums.
  def recent_photos(user_id_or_url, options = {})
    options = make_options(:user_id, user_id_or_url, options)
    options[:recent_photos] = true
    get(options)
  end

  # Retrieves the user's albums and finds the first one with a matching title.
  # Returns a RubyPicasa::Album object.
  def album_by_title(title, options = {})
    if a = user.albums.find { |a| title === a.title }
      a.load options
    end
  end

  # Returns the raw xml from Picasa. See the Picasa.path method for valid
  # options.
  def xml(options = {})
    http = Net::HTTP.new(Picasa.host, 80)
    path = Picasa.path(options)
    response = http.get(path, auth_header)
    if response.code =~ /20[01]/
      response.body
    end
  end

  private

  # If the value parameter is a hash, treat it as the options hash, otherwise
  # insert the value into the hash with the key specified.
  #
  # Uses merge to ensure that a new hash object is returned to prevent caller's
  # has from accidentally being modified.
  def make_options(key, value, options)
    if value.is_a? Hash
      {}.merge value
    else
      options ||= {}
      options.merge(key => value)
    end
  end

  # Combines the cached xml request with the class_from_xml factory. See the
  # Picasa.path method for valid options.
  def get(options = {})
    with_cache(options) do |xml|
      class_from_xml(xml)
    end
  end

  # Returns the header data needed to make AuthSub requests.
  def auth_header
    if token
      { "Authorization" => %{AuthSub token="#{ token }"} }
    else
      {}
    end
  end

  # Caches the raw xml returned from the API. Keyed on request url.
  def with_cache(options)
    path = Picasa.path(options)
    @request_cache.delete(path) if options[:reload]
    xml = nil
    if @request_cache.has_key? path
      xml = @request_cache[path]
    else
      xml = @request_cache[path] = xml(options)
    end
    if xml
      yield xml
    end
  end

  # Returns the first xml element in the document (see
  # Objectify::Xml.first_element) with the xml data types of the feed and first entry
  # element in the document, used to determine which RubyPicasa object should
  # be initialized to handle the data.
  def xml_data(xml)
    if xml = Objectify::Xml.first_element(xml)
      # There is something wrong with Nokogiri xpath/css search with
      # namespaces. If you are searching a document that has namespaces,
      # it's impossible to match any elements in the root xmlns namespace.
      # Matching just on attributes works though.
      feed, entry = xml.search('//*[@term][@scheme]', xml.namespaces)
      feed_scheme = feed['term'] if feed
      entry_scheme = entry['term'] if entry
      [xml, feed_scheme, entry_scheme]
    end
  end

  # Initialize the correct RubyPicasa object depending on the type of feed and
  # entries in the document.
  def class_from_xml(xml)
    xml, feed_scheme, entry_scheme = xml_data(xml)
    if xml
      r = case feed_scheme
      when /#user$/
        case entry_scheme
        when /#album$/
          RubyPicasa::User.new(xml, self)
        when /#photo$/
          RubyPicasa::RecentPhotos.new(xml, self)
        end
      when /#album$/
        case entry_scheme
        when nil, /#photo$/
          RubyPicasa::Album.new(xml, self)
        end
      when /#photo$/
        case entry_scheme
        when /#photo$/
          RubyPicasa::Search.new(xml, self)
        when nil
          RubyPicasa::Photo.new(xml, self)
        end
      end
      if r
        r.session = self
        r
      else
        raise RubyPicasa::PicasaError, "Unknown feed type\n feed:  #{ feed_scheme }\n entry: #{ entry_scheme }"
      end
    end
  end
end
